### PR TITLE
Adds Memory Profiler and CPU Profiler

### PR DESCRIPTION
**Pull request by ishanrai05**

### CODE REVIEW

File: ganga/GangaCore/Utility/Monitoring.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        datafn = mpath+func.__name__ +timestr+".profile"
+        prof = cProfile.Profile()
+        retval = prof.runcall(func, *args, **kwargs)
+        prof.dump_stats(datafn)
+        return retval
+    return wrapper
+
+def cpu_profiler(cls, profile_cpu=c['Profile_CPU']):
+    if profile_cpu:
+        for key, value in list(vars(cls).items()):
```
Comment (alexanderrichards) [inline ganga/GangaCore/Utility/Monitoring.py]: `items()` presumably already returns a list so the explicit list construction seems irrelevant. Also since these are temporary intermediary lists simply used in the iteration perhaps a generator would be better.

i.e. does this not simplify to:
```python
for key, value in vars(cls).iteritems():
```

File: ganga/GangaCore/Utility/Monitoring.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    else:
+        pass
+    return cls
+
+def mem_profiler(cls, profile_memory=c['Profile_Memory']):
+    if profile_memory:
+        cpath = makedir(os.path.join(path,'memory_profiles/'))
+        file_name = cpath+cls.__name__+timestr+'.log'
+        fp = open(file_name, 'w+')
+        for key, value in list(vars(cls).items()):
```
Comment (alexanderrichards) [inline ganga/GangaCore/Utility/Monitoring.py]: ditto

**Review by alexanderrichards — APPROVED** [review]

File: ganga/GangaCore/Utility/Monitoring.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+from GangaCore.Utility.Config import getConfig
```
Comment (mesmith75) [inline ganga/GangaCore/Utility/Monitoring.py]: Is this a new file? Perhaps a better name would be `Profiling`? `Monitoring` makes me think of other things, like the job monitoring.

File: ganga/GangaCore/Utility/Monitoring.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import os
+from memory_profiler import profile
```
Comment (mesmith75) [inline ganga/GangaCore/Utility/Monitoring.py]: I don't think `memory_profiler` is standard. Is the purpose of this branch to go into a release? If so `memory_profiler` needs adding to the requirements.

Comment (egede) [reply ganga/GangaCore/Utility/Monitoring.py]: Yes. To install memory_profiler from pip also requires that the python-devel package is installed on any centos/fedora system. If not already in the centos7 docker image, it will need to be added to  ganga/ganga/GangaCore/test/Dockerfile for the testing to work.

File: ganga/GangaCore/GPIDev/Base/Objects.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
 
 from GangaCore.Utility.Plugin import allPlugins
 
+from GangaCore.Utility.Monitoring import cpu_profiler, mem_profiler
```
Comment (egede) [inline ganga/GangaCore/GPIDev/Base/Objects.py]: Where is the cpu_profiler package from. I did not find it with pip.

File: ganga/GangaCore/Utility/Monitoring.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    if profile_memory:
+        cpath = makedir(os.path.join(path,'memory_profiles/'))
+        file_name = cpath+cls.__name__+timestr+'.log'
+        fp = open(file_name, 'w+')
+        for key, value in list(vars(cls).items()):
+            if callable(value):
+                setattr(cls, key, profile(value, stream=fp, precision=6))
+    else:
+        pass
+    return cls
```
Comment (egede) [inline ganga/GangaCore/Utility/Monitoring.py]: Add the missing newline.

File: ganga/GangaCore/GPIDev/Base/Objects.py [hunk 7]
```diff
@@ -687,6 +689,8 @@ def __init__(cls, name, bases, this_dict):
             this_schema.createDefaultConfig()
 
 
+@cpu_profiler
+@mem_profiler
```
Comment (egede) [inline ganga/GangaCore/GPIDev/Base/Objects.py:694]: Is there a performance price to pay for adding these two, even if the cpu and memory profiler is turned off?

File: ganga/GangaCore/GPIDev/Base/Objects.py (hunk 5 above)
Comment (ishanrai05) [reply ganga/GangaCore/GPIDev/Base/Objects.py]: @egede  cProfile comes pre-installed with python. [docs](https://docs.python.org/2/library/profile.html)

File: ganga/GangaCore/Utility/Monitoring.py (hunk 1 above)
Comment (ishanrai05) [reply ganga/GangaCore/Utility/Monitoring.py]: Yes. Thank you for this.

File: ganga/GangaCore/Utility/Monitoring.py (hunk 3 above)
Comment (ishanrai05) [reply ganga/GangaCore/Utility/Monitoring.py]: Yes, `Profiling` would be a better name.

File: ganga/GangaCore/GPIDev/Base/Objects.py (hunk 7 above)
Comment (ishanrai05) [reply ganga/GangaCore/GPIDev/Base/Objects.py:694]: There is no performance lag on my machine. However, we should test this in real world world scenarios just to be sure.

File: ganga/GangaCore/Utility/Monitoring.py (hunk 6 above)
Comment (ishanrai05) [reply ganga/GangaCore/Utility/Monitoring.py]: I have updated the code to meet PEP 8 requirements.

File: ganga/GangaCore/GPIDev/Base/Objects.py (hunk 7 above)
Comment (alexanderrichards) [reply ganga/GangaCore/GPIDev/Base/Objects.py:694]: No shouldn't really be as these are only run once at module import time and are essentially a no-op when deactivated

File: ganga/GangaCore/Utility/Monitoring.py (hunk 4 above)
Comment (ishanrai05) [reply ganga/GangaCore/Utility/Monitoring.py]: I added  `memory_profiler` to requirements.txt

**Review by egede — APPROVED** [review]
Looks good now.

**Comment by mesmith75** [comment]
Shall we merge this?

**Comment by egede** [comment]
> Shall we merge this?
OK for me. @ishanrai05 what do you say?

**Comment by ishanrai05** [comment]
I added a decorator to count the number of calls made to GangaObject. We can merge this  if you think it is okay.

**Review by egede — APPROVED** [review]
As agreed.
